lib/pbench/cli/agent/commands/pbench_create_user.py

```diff
@@ -0,0 +1,101 @@
+import click
+import requests
+
+from pbench.agent.base import BaseCommand
+from pbench.cli.agent import pass_cli_context
+from pbench.cli.agent.options import common_options
+
+
+class PbenchCreateUser(BaseCommand):
+    def __init__(self, context):
+        super().__init__(context)
+
+    def execute(self):
+        """Creates a user in the pbench user database"""
+
+        server = self.config.agent.parser.get("results", "server_rest_url")
+        headers = {"Content-Type": "application/json", "Accept": "application/json"}
+        payload = {
+            "username": self.context.username,
+            "password": self.context.password,
+            "email": self.context.email,
+            "first_name": self.context.first_name,
+            "last_name": self.context.last_name,
+        }
+        uri = f"{server}/register"
+        try:
+            response = requests.post(uri, headers=headers, json=payload)
+        except requests.exceptions.ConnectionError as exc:
+            raise RuntimeError(f"Cannot connect to '{uri}'") from exc
+
+        if response.status_code == 201:
+            click.echo(f"User {self.context.username} registered")
+            return 0
+
+        if self.context.exists_ok and response.status_code == 403:
+            click.echo(f"User {self.context.username} already registered")
+            return 0
```

Comment:
In terms of transient hackery, `exists_ok` is about the _gold_ unit tests and test isolation, with the idea that a test's `.db` companion file can specify users and the `unittests` script will create them. One option is for the test script to start out creating a certain set of users and unit tests are required to use (only) those; however as a completely isolated per-test setup, the "create if doesn't exist" made more sense. I suppose another option would be for the script to do `pbench-create-user` (which reminds me, all the other command names are `-` separated; this one is `_` separated, and should probably be changed) and have it simply return the HTTP status for errors and the script would ignore an "already exists" error.

`pbench-move-results` is definitely an agent command, as that's where performance results are generated, and it doesn't have much meaning elsewhere. I suppose that `pbench-generate-token` is potentially more nuanced, particularly when we start supporting token roles. It could hardcode a "PUT-only" token, or it could be evolved with an option to specify the token role in which case it could be used to generate a token for user delete/modify by the owner or admin.

In theory, `pbench-create-user` could be a convenience if someone is generating data and hasn't already gone to the UI to register; I'm not sure that's such a big deal at this point, but it's a possibility.

General note in terms of naming ... if we're going to move to a single master command with sub-commands, like `git`, it might make more sense to adopt going forward a __pbench__-_verb_-_noun_ naming style that will collapse more naturally into that schema. So maybe we want `pbench-user-create` and `pbench-token-generate` and even `pbench-results-move`? Of course that changes most all of our command names, which might be excessive...

Reply:
> In terms of transient hackery

Is it transient?  Do you see the need for this (e.g., in the unit tests) as going away sometime soon?  Perhaps I'm naive, but I expect that we'll be testing this functionality pretty much indefinitely.

> all the other command names are - separated; this one is _ separated, and should probably be changed

Yes, it should certainly be changed!  My understanding of the Click support is that, once everything is in place, we'll be able to invoke commands either with hyphens (`pbench-user-create`) or spaces (as subcommands, `pbench user create`).  Underscores would be parsed completely differently.

> another option would be for the script to do pbench-create-user [...] the script would ignore an "already exists" error.

That would certainly be my preference:  that keeps the tool more focused and better separates policy from mechanism.

> `pbench-move-results` is definitely an agent command

:-)  Spoken like a server-side guy.  :-D

> I suppose that `pbench-generate-token` is potentially more nuanced

Yes, I'm expecting that, given our seeming hunger for command-line tools, it's applicability is only going to grow.

> In theory, pbench-create-user could be a convenience if someone is generating data and hasn't already gone to the UI to register; I'm not sure that's such a big deal at this point, but it's a possibility.

I would be inclined to push back on that:  for human users, creating a username is a one-time task, the first step toward examining data (they may or may not decide to generate their own data to examine) -- if they don't eventually use the GUI (to set up sharing, if they don't view their data themselves), then there is no point to uploading any data, so they might as well fire up the GUI first and create their username.

> if we're going to move to a single master command with sub-commands, like git, it might make more sense to adopt going forward a pbench-verb-noun naming style

Yes, I concur.  (I should have done that with `pbench-generate-token`, but I followed the existing practice instead.)

Reply:
> But, that leads to a couple more questions. First, how do we know that a `403` response means "user already exists"? I'm not even sure that we should be seeing a `403`, because that status implies that the requestor has been authenticated, and this request doesn't involve that. I realize that we don't want to leak or otherwise create a way for an assailant to probe for existing users, but I'm not sure how to avoid it and still enable the `exists_ok` concept. (Now, if the request were coming from an authenticated `admin` user on behalf of a third party, life would be very different!). And, my other question centers around the fact that "already exists" implies/suggests that the existing user _matches_ the one in the creation request (same first/last name and email, in addition to the same username); so, really, before deciding whether the `403` is an error or not, we would need to fetch the existing information for the requested user, and decide whether it matches (which, of course, would require authentication, either using the supplied credentials or by running this thing as an admin in the first place).

There is no authentication layer involved when registering a user, meaning a user doesn't need to provide any auth token in the request. So at the moment there is no way to know if the requestor is an admin or someone else when they are just registering a user. We know `403` means a user already exists because thats the status code we defined for user already exist conditions (with varying messages such as duplicate username or email, because these are the only 2 unique constraints in our db). To fetch a user information the request needs pbench authentication token, which is not a required header for create user api. What I can do is I think should just return the message the api returned.

> And, that leads into a discussion touching on whether this is a "server-side" tool, an "agent-side" tool, or something else entirely. If it's a server-side tool, I'm not sure that it needs to be using the client REST interface -- it could be using some "internal" channel which we can trust without too much authentication. On the other hand, it doesn't seem like it makes sense as an agent-side tool -- maybe this is my lack of understanding -- but it doesn't seem like the _agent_ should be involved with things like user management -- that's a server concept used by clients (like the Dashboard and `pbench-move-results` which, admittedly, is an agent tool, but it's interfacing with the server to perform a user function). So, it feels like the proposed tool belongs to a new category..."client-tools". (And, presumably, `pbench-generate-token` should move into this category, and possibly `pbench-move-results`, as well.) And, once we figure out which category it belongs in, then we can figure out how to configure it to get access to the server.

The reason I choose this to be an agent side tool is because on the spectrum of agent -> server I felt its slightly more closer to agent side since most of the time the tool is more useful and convenient from outside the server machine to create a user via cli. (I considered anything outside server is agent but I agree client is more suitable terminology) We dont have the client category yet. 

And as @dbutenhof mentioned the `exists_ok` option is really for the gold unit tests and test isolation. It just ignores if there is an already existing `username` or `email` (403) and returns success without actually creating a new user. Every other error is reported back to the user.

@dbutenhof should I move this utility under server side?

Reply:
> There is no authentication layer involved when registering a user, meaning a user doesn't need to provide any auth token in the request. So at the moment there is no way to know if the requestor is an admin or someone else when they are just registering a user. We know 403 means a user already exists because thats the status code we defined for user already exist conditions (with varying messages such as duplicate username or email, because these are the only 2 unique constraints in our db). To fetch a user information the request needs pbench authentication token, which is not a required header for create user api. What I can do is I think should just return the message the api returned.

I agree.  So, there are various issues (some of which are or may be outside the scope of this PR):

- As currently constructed, the `/register` API endpoint leaks security information which could potentially be abused (i.e., it would allow an attacker to probe for existent usernames).  Is this or should this be a concern?  (Should we be returning something other than `403` and/or should we be less explicit with our error messages?)
- Do we want to continue to offer self-registration, or should we switch to a model which requires users to be created by a suitably-privileged, existent account?  (This creates annoying problems of its own, but most of them go away once we integrate with an external authorization facility such as LDAP or RH SSO, and it delivers us from having to cope with unauthenticated accesses.)
- For this PR, we _could_ consider validating the user-exists scenario by trying to use the supplied credentials to execute a login:  if it works, then, yes, indeed, the user does already exists and we're good to go; if it doesn't, then the user that they are trying to create already exists in an incompatible way, things aren't going to work out well, and we should report an error.

> The reason I choose this to be an agent side tool is because on the spectrum of agent -> server I felt its slightly more closer to agent side since most of the time the tool is more useful and convenient from outside the server machine to create a user via cli. (I considered anything outside server is agent but I agree client is more suitable terminology) We dont have the client category yet.

Again, I agree.  In the grand scheme of things, it seems much more likely that end-users will have access to a machine with the agent installed than that they will have access to a machine with the server installed.  So, if we want this tool to be useful to the end-user, it should not be a server-side tool.  However, what we want this tool for initially is for testing the server...which means that either it should be a server-side tool (not exactly desirable), or we need to have the agent available when we test the server (pretty much undesirable), or we need to get on with creating a third category (better than "not exactly desirable"?).

> It just ignores if there is an already existing username or email (403) and returns success without actually creating a new user.

Understood.  However, this could be dealt with in the testing setup instead of in the tool....

Reply:
> * As currently constructed, the `/register` API endpoint leaks security information which could potentially be abused (i.e., it would allow an attacker to probe for existent usernames).  Is this or should this be a concern?  (Should we be returning something other than `403` and/or should we be less explicit with our error messages?)

Yes I guess we can be little less explicit with our return messages, however, idea behind these return messages like (`username/email already in use`) is to just give some help to the user to understand what went wrong and then he/she can correct the request with appropriate values. Without any error/help the user would keep guessing what might have went wrong and after couple of failed attempts the user might just leave the room and not come back (we dont want that to happen, we do want user to quickly amend the request without much hassle)


> * Do we want to continue to offer self-registration, or should we switch to a model which requires users to be created by a suitably-privileged, existent account?  (This creates annoying problems of its own, but most of them go away once we integrate with an external authorization facility such as LDAP or RH SSO, and it delivers us from having to cope with unauthenticated accesses.)

This needs more discussion on how to do it and explore our options before I can dive into it. There are certainly few ways to write SSO login scripts with flask and that including all the major SSO providers (gmail, github, redhat etc). I guess we can't just rely on RH SSO since other people who do not have redhat account might want to use the product. However, we certainly need our user database as well.


> * For this PR, we _could_ consider validating the user-exists scenario by trying to use the supplied credentials to execute a login:  if it works, then, yes, indeed, the user does already exist and we're good to go; if it doesn't, then the user that they are trying to create already exists in an incompatible way, things aren't going to work out well, and we should report an error.

I am not sure but I do think this could be an overkill for such a simple utility tool, where we have to hit two rest endpoints in order to create a user that is login and logout, if we login we create pbench token in our database and then logout to remove that token (Potentially 2 database queries). Since this is a server side utility it is assumed that whoever wants to use the cli to create user should have server installed already and so we could just simply pass on the error message that rest endpoint reported.

Reply:
Chained conversations in GitHub are awkward to unwind!

> Is it transient? Do you see the need for this (e.g., in the unit tests) as going away sometime soon? Perhaps I'm naive, but I expect that we'll be testing this functionality pretty much indefinitely.

@webbnh  I see the "gold" CLI-level unit tests going away, which would remove the need for this. Not right away, unfortunately.

> The reason I choose this to be an agent side tool is because on the spectrum of agent -> server I felt its slightly more closer to agent side since most of the time the tool is more useful and convenient from outside the server machine to create a user via cli. (I considered anything outside server is agent but I agree client is more suitable terminology) We dont have the client category yet.

@npalaska As @webbnh commented, registration is a one-time occurrence per-user and I just don't see how it makes much sense to offer this as part of the agent RPM/container. And, more immediately, since the sense of urgency that caused me to push you on this was primarily motivated by the need to create users for the CLI unit tests in #2101 ... this PR, as written, does me no good at all as the server unit tests can't depend on an agent CLI environment!

Plus, again, I'd originally conceived this as a generic user management CLI tool primarily for an administrator to use as part of the server installation.

However, I can definitely see some benefit to Webb's vision of a separate "client" package that might be compatible with either server or agent, based on HTTP/S communication with an installed server... that could certainly include token generation, user management, and possibly also move/copy results. (The later based on the argument that using the agent for benchmark orchestration and results consolidation and local analysis could be a sensible use case without any interaction at all with a Pbench server, and move/copy results is only necessary if you _are_ using a server.)

(Although in practice I'm not sure we want to require installing a second package to enable this, or go to the effort right now to create such a package.)

> Yes I guess we can be little less explicit with our return messages, however, idea behind these return messages like (username/email already in use) is to just give some help to the user to understand what went wrong and then he/she can correct the request with appropriate values. Without any error/help the user would keep guessing what might have went wrong and after couple of failed attempts the user might just leave the room and not come back (we dont want that to happen, we do want user to quickly amend the request without much hassle)

@npalaska This whole topic get complicated; so many tradeoffs to consider. Some web sites I've used support "self registration" but don't allow you to just enter your info and submit... instead, you start with an email; it sends you a registration link to ensure that your email is valid, and presumably has already checked that there's no existing account with that email. The email link gets you to a secondary registration page where you get to choose a username and password now that the email has been validated. This sort of scheme makes it a bit harder to deploy an account sniffer, without requiring an admin to intercede for every account request. There are plenty of other ideas out there, too.

Maybe for now a simple 400 error for any sort of error would at least make the job a bit harder for a sniffer? Neither 401 nor 403 are really appropriate here, anyway, since no authorization is involved in the first place.

Reply:
> idea behind these return messages like (username/email already in use) is to just give some help to the user to understand what went wrong

Nikhil, I'm right there with you:  I prefer errors to be specific and clear, because, at least initially, _I_ will be the user, and I'll appreciate knowing what's going on when something goes wrong.  But, in the case of reporting back to an unknown, un-credentialed, unauthorized request, we are supposed to be very circumspect about what we give away.  Generally speaking, it's supposed to be "one bit of resolution" -- success or failure, and not much more.  @dbutenhof's suggestion of using `400` is probably appropriate (I don't like it much, but I think it's correct).  If we want to support the user better, then they need to log in first so that we can account for their activities and our responses, and that gets us out of the area of self-registration.

> after couple of failed attempts the user might just leave the room and not come back

I think you're right to be sensitive to this.  However, to support that sort of user, we can use the GUI, and we can implement multiple stages to registration, as Dave outlined.  For the case in this PR, I'm not worried about the user walking away, because I expect that someone using a command line tool to create a registration has an intense interest in making it succeed, and they probably aren't intending to register only a single time (for that, the GUI is likely an easier approach for the user).

> I guess we can't just rely on RH SSO since other people who do not have redhat account might want to use the product.

I think that that depends on the deployment.  For instance, if we are offering our service Red Hat-wide, then RH SSO would probably be just the thing.  For instance, Quay is adopting this model, supporting both internal and external users via RH SSO, with no alternative authentication mechanism.  Obviously, if we want to support other deployments, we would presumably have to support other authentication mechanisms, as well.

> I am not sure but I do think this could be an overkill for such a simple utility tool, where we have to hit two rest endpoints in order to create a user (login and logout), if we login we create pbench token in our database and then logout to remove that token (Potentially 2 database queries).

I'm assuming that it would actually be as many as three endpoints:  attempt to create; when that fails, attempt to log in; if that succeeds, log out.  In human-scale terms, I don't think the user will be able to distinguish one request from three (think about how many requests the browser has to make to load the typical, modern, JS-based web page!).  And, it only makes multiple requests in the create-failure case (and it only makes the third request in the true-duplicate case).  But, this would allow the tool to securely determine whether the request is for a duplicate username.

As for whether it is overkill, yeah, it probably is.  In particular, it's not doing anything that the calling script couldn't do.  That is, the calling script could attempt the creation; if it fails, it could attempt to generate a token; if _that_ fails then there is a real error; otherwise, the failed creation is a non-error, and we're good to go.  And, all this works with just binary answers from the tools.

> Since this is a server side utility it is assumed that whoever wants to use the cli to create user should have server installed already and so we could just simply pass on the error message that rest endpoint reported.

The issue is not so much the behavior of this tool but rather the behaviors of the server required to support this tool.  That is, any REST API which this tool can use _could_ also be used by a piece of malicious software sitting elsewhere on the network.  So, if the server is giving effusive responses to allow this tool to support its user, then the attacker is also going to have that same level of help and support.

I would be inclined to pursue the idea that this tool is a server-side utility and therefore is co-located with the server.  That would enable a _private_ communication with the server which could be trusted without authentication.  For instance, we could provide a listener on a local port which doesn't accept connections from other hosts.  Or, we could give the REST endpoint different behaviors (e.g., different diagnostics, different capabilities, etc.) depending on whether the request comes from the local host or from a remote host.  If we know from context that we can trust the requestor, then lots of these issues go away.

Reply:
> I see the "gold" CLI-level unit tests going away, which would remove the need for this. Not right away, unfortunately.

@dbutenhof, understood.  However, once the "gold" unit tests are gone, won't we still have unit tests which need to have usernames in order to perform their function?  Or, would we do it all with mocks?  (Still, for _functional_ testing, we would want to be able to have usernames, and being able to create them automatically for the tests would be handy, so I don't really see the need for this tool going away.)

> in practice I'm not sure we want to require installing a second package to enable this, or go to the effort right now to create such a package.

Don't bother me with trivialities!  :-D

(You may well be right, though; still, we have to know where we want to go before we can decide how and when we want to go there.)